src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -0,0 +1,150 @@
+
+                        status = Interop.Crypto.CmsAddSkidRecipients(cms, skidRecipientCerts);
+                        CheckStatus(status);
+
+                        status = Interop.Crypto.CmsAddIssuerAndSerialRecipients(cms, issuerRecipientCerts);
+                        CheckStatus(status);
+
+                        status = Interop.Crypto.CmsAddOriginatorCerts(cms, originatorCertStack);
+                        CheckStatus(status);
+                        
+                        Interop.Crypto.BioWrite(contentBio, contentInfo.Content, contentInfo.Content.Length);
+                        status = Interop.Crypto.CmsEncryptFinalize(cms, contentBio);
+                        CheckStatus(status);
+
+                        byte[] encryptedMessage = Interop.Crypto.OpenSslEncode(
+                            Interop.Crypto.CmsGetEncodingLength,
+                            Interop.Crypto.CmsGetEncoding,
+                            cms);
+
+                        // TODO(3334): Add support for unprotected attributes here.
+
+                        return encryptedMessage;
+                    }
+                }
+            }
+        }
+
+        private void CheckStatus(int status)
+        {
+            switch(status)
+            {
+                case -1:
+                    // We should never reach this state. It checks for null inputs given to shim functions, but as we
+                    // always call CheckValidOpenSslHandle on a handle before using it there should never be such problem.
+                    throw new ArgumentNullException();
+
+                case 0:
+                    throw Interop.Crypto.CreateOpenSslCryptographicException();
+
+                case 1:
+                    // This is the expected state.
+                    return;
+            }
+            System.Diagnostics.Debug.Fail("Unxepected status from shim. Check return values");
+        }
+
+        private void ClassifyRecipients(CmsRecipientCollection recipients, SafeX509StackHandle issuerRecipientCerts, SafeX509StackHandle skidRecipientCerts)
+        {
+            foreach (CmsRecipient recipient in recipients)
+            {
+                // This shouldn't happen as the CmsRecipient constructor throws an exception when a null recipient is passed,
+                // but for desktop compat in case someone manages to change the certificate, this is the exception Framework throws.
+                if (recipient.Certificate == null)
+                    throw new ArgumentNullException(SR.Cryptography_Cms_RecipientCertificate_Not_Found);
+
+                // We cannot encrypt KeyAgreement now as there is no native support for it in OpenSSL and it would require manual parsing
+                // of the data, so we have to make sure it's KeyTransport.
+                if (recipient.Certificate.GetKeyAlgorithm() != Oids.Rsa)
+                    throw new PlatformNotSupportedException();
```

Comment:
This should have a message.  Something like

Cryptography_Cms_KeyAgreementPlatformNotSupported=This platform does not support encrypting for KeyAgreement recipients.